src/main/scala/smack/messaging/Client.scala

```diff
@@ -114,60 +155,95 @@ class Client extends FSM[State, Context] {
+      chat.channel.sendMessage(messageToSend)
       log.info(s"message sent to $recipient")
-      stay using ctx.copy(chats = chats + (recipient → chat))
-
-    case Event(Messages.SendFileMessage(recipient, fileUrl, description), ctx) ⇒
-      val (user, domain) = splitUserIntoNameAndDomain(recipient)
-      val chat = ctx.chats.getOrElse(key = recipient, createChat(ctx.connection.get, user, domain))
-      val fileInformation = OutOfBandData(fileUrl, description)
+      sender ! MessageId(messageToSend.getStanzaId)
+      val msgState = MessageState(message, MessageId(messageToSend.getStanzaId), Unacknowledged)
+      stay using ctx.copy(chats = ctx.chats + (fullUser → chat.copy(messages = chat.messages :+ msgState)))
+
+    case Event(Messages.SendUrlMessage(recipient, fileUri, description), ctx) ⇒
+      val (user, domain) = recipient.splitUserIntoNameAndDomain(defaultDomain)
+      val fullUser = user getFullyQualifiedUser domain
+      val chat = ctx.chats.getOrElse(key = fullUser, ChatState(createChat(ctx.connection.get, user, domain), Seq.empty))
+      val fileInformation = OutOfBandData(fileUri, description)
       val infoText = "This message contains a link to a file, your client needs to " +
         "implement XEP-0066. If you don't see the file, kindly ask the client developer."
       val message = new Message(recipient.value, infoText)
       message.addExtension(fileInformation)
-      chat.sendMessage(message)
+      chat.channel.sendMessage(message)
       log.info(s"file message sent to $recipient")
-      stay using ctx.copy(chats = ctx.chats + (recipient → chat))
+      sender ! MessageId(message.getStanzaId)
+      val msgState = MessageState(message.getBody, MessageId(message.getStanzaId), Unacknowledged)
+      stay using ctx.copy(chats = ctx.chats + (fullUser → chat.copy(messages = chat.messages :+ msgState)))
+
+    case Event(Messages.SendFileMessage(recipient, file, description), ctx) ⇒
+      val uploadFuture = uploadAdapter.upload(file, description)
+      uploadFuture onComplete {
+        case Success(uri) ⇒
+          self ! Messages.FileUploaded(User(ctx.connection.get.getUser), uri, description)
+          self ! Messages.SendUrlMessage(recipient, uri, description)
+        case Failure(ex) ⇒
+          log.error(ex, s"could not upload file!")
+          self ! Messages.FileUploadError(ex)
+      }
+      stay
 
     case Event(register: Messages.RegisterUser, Context(Some(connection), chats, _)) ⇒
       log.info(s"trying to register ${register.user}")
       val accountManager = AccountManager.getInstance(connection)
       Try {
-        val (username, _) = splitUserIntoNameAndDomain(register.user)
-        if (username.value == domain) throw new Messages.InvalidUserName(register.user)
+        val (username, _) = register.user.splitUserIntoNameAndDomain(defaultDomain)
+        if (username.value == defaultDomain.value) throw new Messages.InvalidUserName(register.user)
```

Comment:
Doesn't the try block catch it and then it matches on Failure down below, so it is properly wrapped up in a Failure? That is what the test shows at least -- and I feel it makes the intent more clear because then all exceptions are handled in one block regardless of where it comes from; instead of having it be separate